### PR TITLE
Add comprehensive memory model tests

### PR DESCRIPTION
**Pull request by reeselevine**
Issue: #963 

- Adds remaining memory model tests from https://gpuharbor.ucsc.edu/webgpu-mem-testing/conformance/
- A total of 40 memory model tests are now included. On my Intel Macbook, the tests finish in ~30 seconds.
- A couple minor updates to the setup code are included so that all tests work correctly and fixed parameters can't be accidentally changed
- Parameter sets for coherence/weak memory tests are updated to use values found to be the most effective when doing experiments

@dneto0 @alan-baker @tyler-utah

P.S. I'm not sure if it's unexpected that `package-lock.json` has changes. If this is bad, I can revert it to the current `package-lock.json` on the main branch.
<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -52,12 +52,13 @@
       }
```
Comment (zoddicus) [inline package-lock.json:52]: I wouldn't expect this needed to be updated unless you are explicitly using functionality from newer packages, so I would probably revert to main unless there is something specific you need to update for.

**Review by zoddicus — COMMENTED** [review]
I have added a review comment about the package-lock.json change, but scanning over the rest of the PR, I don't think I am qualified to approve it in general, since I am not particularly familiar with the memory model and workgroup management in WGSL. Since @dneto0 is away I have added @ben-clayton who hopefully will have enough context to give a proper review.

**Comment by dneto0** [comment]
I plan to review this.  It's in my queue.

File: src/webgpu/shader/execution/memory_model/coherence.spec.ts [hunk 2]
```diff
@@ -54,6 +52,22 @@ const storageMemoryCorrTestCode = `
...
+  let r0 = atomicLoad(&test_locations.value[x_1]);
+  let r1 = atomicLoad(&test_locations.value[y_1]);
+  atomicStore(&results.value[shuffled_workgroup * workgroupXSize + id_1].r0, r0);
+  atomicStore(&results.value[shuffled_workgroup * workgroupXSize + id_1].r1, r1);
+`;
+
+const corrRMWTestCode = `
+  atomicExchange(&test_locations.value[x_0], 1u);
+  let r0 = atomicLoad(&test_locations.value[x_1]);
+  let r1 = atomicAdd(&test_locations.value[y_1], 0u);
```
Comment (dneto0) [inline src/webgpu/shader/execution/memory_model/coherence.spec.ts:66]: So we have two independent dimensions:    storage vs. workgroup; and  read-read vs read-RMW.
So there should be 4 variants, but there are only 3.

Seems we should also test RMW on the workgroup storage.

Comment (dneto0) [reply src/webgpu/shader/execution/memory_model/coherence.spec.ts:66]: For consistency with the others, should this be called storageMemoryCorrRMWTestCode ?

File: src/webgpu/shader/execution/memory_model/coherence.spec.ts [hunk 3]
```diff
@@ -98,18 +118,305 @@ g.test('corr')
...
+    );
+    await memModelTester.run(60, 3);
+  });
+
+const storageMemoryCowwTestCode = `
+  atomicStore(&test_locations.value[x_0], 1u);
+  atomicStore(&test_locations.value[y_0], 2u);
+`;
 
+const cowwRMWTestCode = `
```
Comment (dneto0) [inline src/webgpu/shader/execution/memory_model/coherence.spec.ts:137]: suggest renaming to storageMemoryCowwRMWTestCode

File: src/webgpu/shader/execution/memory_model/coherence.spec.ts [hunk 4]
```diff
@@ -98,18 +118,305 @@ g.test('corr')
...
+  atomicExchange(&test_locations.value[x_0], 1u);
+  atomicStore(&test_locations.value[y_0], 2u);
+`;
+
+const workgroupMemoryCowwTestCode = `
```
Comment (dneto0) [inline src/webgpu/shader/execution/memory_model/coherence.spec.ts:142]: Suggest adding a RMW variant too.

File: src/webgpu/shader/execution/memory_model/coherence.spec.ts [hunk 5]
```diff
@@ -98,18 +118,305 @@ g.test('corr')
...
+`;
+
+const workgroupStorageMemoryCowrTestCode = `
+  atomicStore(&test_locations.value[x_0], 1u);
+  let r0 = atomicLoad(&test_locations.value[y_0]);
+  atomicStore(&test_locations.value[x_1], 2u);
+  atomicStore(&results.value[shuffled_workgroup * workgroupXSize + id_0].r0, r0);
+`;
+
+const cowrRMWTestCode = `
```
Comment (dneto0) [inline src/webgpu/shader/execution/memory_model/coherence.spec.ts:211]: suggest remaing to storageCowrRMWTestCode

File: src/webgpu/shader/execution/memory_model/coherence.spec.ts [hunk 6]
```diff
@@ -98,18 +118,305 @@ g.test('corr')
...
+  atomicExchange(&test_locations.value[x_0], 1u);
+  let r0 = atomicAdd(&test_locations.value[y_0], 0u);
+  atomicExchange(&test_locations.value[x_1], 2u);
+  atomicStore(&results.value[id_0].r0, r0);
+`;
+
+const workgroupMemoryCowrTestCode = `
```
Comment (dneto0) [inline src/webgpu/shader/execution/memory_model/coherence.spec.ts:218]: Suggest adding a RMW variant

File: src/webgpu/shader/execution/memory_model/coherence.spec.ts [hunk 7]
```diff
@@ -98,18 +118,305 @@ g.test('corr')
...
+`;
+
+const workgroupStorageMemoryCorw2TestCode = `
+  let r0 = atomicLoad(&test_locations.value[x_0]);
+  atomicStore(&test_locations.value[y_0], 1u);
+  atomicStore(&test_locations.value[x_1], 2u);
+  atomicStore(&results.value[shuffled_workgroup * workgroupXSize + id_0].r0, r0);
+`;
+
+const corw2RMWTestCode = `
```
Comment (dneto0) [inline src/webgpu/shader/execution/memory_model/coherence.spec.ts:354]: similarly, rename

File: src/webgpu/shader/execution/memory_model/coherence.spec.ts [hunk 8]
```diff
@@ -98,18 +118,305 @@ g.test('corr')
...
+  let r0 = atomicLoad(&test_locations.value[x_0]);
+  atomicStore(&test_locations.value[y_0], 1u);
+  atomicExchange(&test_locations.value[x_1], 2u);
+  atomicStore(&results.value[id_0].r0, r0);
+`;
+
+const workgroupMemoryCorw2TestCode = `
```
Comment (dneto0) [inline src/webgpu/shader/execution/memory_model/coherence.spec.ts:361]: similarly, add a RMW variant for workgroup memory

File: src/webgpu/shader/execution/memory_model/memory_model_setup.ts [hunk 9]
```diff
@@ -959,12 +983,15 @@ export function buildTestShader(
   testType: TestType
 ): string {
   let memoryTypeCode;
+  let isStorageClass = false;
```
Comment (dneto0) [inline src/webgpu/shader/execution/memory_model/memory_model_setup.ts:986]: FYI.  We renamed "storage class" to "address space", so you can rename this  isStorageAS or something like that.

**Review by dneto0 — CHANGES_REQUESTED** [review]
Looking really good.
Two requests:
- more consistent naming of the test code snippets.
- add RMW workgroup-storage variants when there are storag-buffer variants.

**Comment by reeselevine** [comment]
Added more RMW variants! There are actually three "types" of coherence tests: inter-workgroup on storage address space, intra-workgroup on storage as, and intra-workgroup on workgroup as. I've added rmw variants for all three of these, as well as addressed the other minor comments.

**Review by dneto0 — APPROVED** [review]
I understand even better now.
LGTM, and thanks!

**Pull request merged**